gemrb/core/GameData.cpp

```diff
@@ -239,55 +243,31 @@ bool GameData::DelTable(unsigned int index)
 	return true;
 }
 
-Palette *GameData::GetPalette(const ieResRef resname)
+PaletteHolder GameData::GetPalette(const ResRef resname)
 {
-	Palette *palette = (Palette *) PaletteCache.GetResource(resname);
-	if (palette) {
-		return palette;
-	}
-	//additional hack for allowing NULL's
-	if (PaletteCache.RefCount(resname)!=-1) {
-		return NULL;
-	}
+	auto iter = PaletteCache.find(resname);
+	if (iter != PaletteCache.end())
+		return iter->second;
+
 	ResourceHolder<ImageMgr> im = GetResourceHolder<ImageMgr>(resname);
 	if (im == nullptr) {
-		PaletteCache.SetAt(resname, NULL);
+		PaletteCache[resname] = NULL;
 		return NULL;
 	}
 
-	palette = new Palette();
+	PaletteHolder palette = new Palette();
 	im->GetPalette(256,palette->col);
 	palette->named=true;
-	PaletteCache.SetAt(resname, (void *) palette);
+	PaletteCache[resname] = palette;
 	return palette;
 }
 
-void GameData::FreePalette(Palette *&pal, const ieResRef name)
+void GameData::FreePalette(PaletteHolder &pal, const ieResRef)
 {
-	int res;
-
-	if (!pal) {
-		return;
-	}
-	if (!name || !name[0]) {
-		if(pal->named) {
-			error("GameData", "Palette is supposed to be named, but got no name!\n");
-		} else {
-			pal->release();
-			pal=NULL;
-		}
-		return;
-	}
-	if (!pal->named) {
-		error("GameData", "Unnamed palette, it should be %s!\n", name);
-	}
-	res=PaletteCache.DecRef((void *) pal, name, true);
-	if (res<0) {
-		error("Core", "Corrupted Palette cache encountered (reference count went below zero), Palette name is: %.8s\n", name);
-	}
-	if (!res) {
-		pal->release();
-	}
+	// This was previously much hairier, trying to keep track of two different
+	// palette refcounts.  Now we just rely on Holder/Held to make sure memory
+	// is freed, while not bothering about freeing named palettes from the
+	// map.
 	pal = NULL;
```

Comment:
are "named" `Palettes` even a thing anymore? this function is just clearing a `PaletteHolder` and is no different from just assigning `nullptr` to the holder. I would prefer to just do that and nix this function.

Reply:
That's not something I can answer. To me it looks like palettes can be looked up by name, making them named, and then also copied into unnamed ones, or just created unnamed. I suspect the Cache here was trying to do two things, keep a cache of named palettes, and also provide refcounting for other palettes. But it was all quite opaque.

I'll commit to making another pass to remove this function in a followup patch (possibly after tackling sprites). Ok to leave as is for now?

Reply:
sure, thats fine.